### DISCUSSION TITLE
Why on onDeviceDiscovered the type is BluetoothDeviceEvent but when print the result is different?

### DISCUSSION THREAD

**Discussion by ValentinOsvaldo** [kenjdavidson/react-native-bluetooth-classic#262] [General]
Hello

I have tried to use the ```onDeviceDiscovered``` function to save the devices as it finds them, but the type of the parameter returned by the callback is BluetoothDeviceEvent but when I print the result to the console, it returns something like this:

```js
{
  "address": "example",
  "deviceClass": {
    "majorClass": 256,
    "deviceClass": 256
  },
  "extra": {
    "rssi": -81
  },
  "bonded": false,
  "id": "example",
  "name": "Totalplay"
}
```

I notice that there is an interface called ```BluetoothNativeDevice```, very similar to what the ```onDeviceDiscovered``` listener prints. Is this really the type of value that it returns as a parameter in the listener?

As another question, I see that in types like ```BluetoothDevice``` there is the rssi as the value of the interface, but it is really printed normally within the value of extra. Is there something that makes it change?

![image](https://github.com/kenjdavidson/react-native-bluetooth-classic/assets/69062379/1d6ab01b-62d7-4af4-b534-a0ed438efb30)

Thanks for your help :D

**Comment by kenjdavidson** [comment]
1.  You are correct, it looks like https://github.com/kenjdavidson/react-native-bluetooth-classic/blob/f47d8ff1d979321606f7474ec472319ba6d244c4/android/src/main/java/kjd/reactnative/bluetooth/receiver/DiscoveryReceiver.java#L52 is actually creating a `BluetoothNativeDevice`, looks like this is a typing error.   In the interim you can:

a) Attempt to just use `event as BluetoothNativeDevice` and see if it works.
b) Fix the issue and open a PR for it.

The response from that should actually be, something like:

```
BluetoothDeviceEvent: {
  eventType: BLUETOOTH_DEVICE,
  device: {
     address: ...
     etc...
  }
}
```

To match the rest of the event data.

3.  This looks like another issue, the `rssi` is only in the `extra` map, I don't think it's ever placed within the main object.   Although I don't think it really matters.  For backward compatibility the line https://github.com/kenjdavidson/react-native-bluetooth-classic/blob/f47d8ff1d979321606f7474ec472319ba6d244c4/android/src/main/java/kjd/reactnative/bluetooth/receiver/DiscoveryReceiver.java#L53C23-L53C23 can be updated to set it in both places.

It would also need to be added to the Java version of `NativeDevice` in order to do so.

**Comment by ValentinOsvaldo** [comment]
I want to create a PR but to fix this typed issue there are to modify the https://github.com/kenjdavidson/react-native-bluetooth-classic/blob/main/src/BluetoothModule.ts#L484?

**Reply by kenjdavidson** [reply]
That line just returns the created subscription, it doesn't really do much with regards to the object returned from the subscription listener/handler.

**Comment by ValentinOsvaldo** [comment]
While I was changing my code I decided to try some use cases like connecting while the device is already connected but I noticed that it gives me this error

![image](https://github.com/kenjdavidson/react-native-bluetooth-classic/assets/69062379/79a09fc3-3a07-4534-bc4c-27d208dd97e5)0

This is my code:

```ts
const connectDevice = (device: BluetoothNativeDevice): Promise<void> => {
    return new Promise(async (resolve, reject) => {
      try {
        const connectionSuccess = await RNBluetoothClassic.connectToDevice(
          device.address,
          {
            connectionType: 'binary',
            delimiter: '\r\n',
          },
        );

        if (!connectionSuccess)
          throw new Error(`Error al intentar conectar ${device.name}`);

        const readDevice = RNBluetoothClassic.onDeviceRead(
          device.address,
          onReceiveData,
        );

        const onDisconnect = RNBluetoothClassic.onDeviceDisconnected(
          (event: any) => {
            readDevice.remove();
            onDisconnect.remove();

            dispatch(removeDevice(event.address));

            ToastAndroid.show(`${event.name} se ha desconectado`, 3000);
          },
        );

        dispatch(
          addListener({
            address: device.address,
            listener: readDevice,
          }),
        );

        dispatch(setConnectedDevice(device));
        resolve();
      } catch (error) {
        reject(error);
      }
    });
  };
```

The ```connectToDevice``` method return a object, BluetoothDevice's connect method returns either true or false. How would I catch these exceptions?

**Reply by kenjdavidson** [reply]
Your screenshot and question are a little confusing:
- The screenshot shows an exception being thrown while attempting to set `onReadDevice` to a device which the native module doesn't have in it's connected devices collection.  The error should have been caught in your `catch(error)` at the bottom, since you are wrapping the `onReadDevice` within that `try/catch` block.
- The reason behind this is that `device.connect()` you already have the device, so there is no reason to return the `BluetoothDevice` from that, a true/false seems acceptable.   `RNBluetoothClassic.connectToDevice(address)` needs to return the full `BluetoothDevice` so that you can start working with it (if you choose) or you can choose not to and just keep the `address` for use.   I tried to leave the option to use static methods or instance methods up to the developer.

I would say use the appropriate method for what you want to do?

**Reply by ValentinOsvaldo** [reply]
Sorry if my question is confusing...

My problem is the error on the screenshot doesn't catch the error because the ```connectToDevice``` method returns a object, I mean the request is success, so I want to catch the error but the object does not say anything to catch the error (in case the device is connected to another). In the ```connect``` method of ```BluetoothDevice``` returns a boolean, that helps me detect whether to continue with the process or not.

**Reply by ValentinOsvaldo** [reply]
I find other way to prevent the error:

```ts
await RNBluetoothClassic.connectToDevice(device.address, {
  connectionType: 'binary',
  delimiter: '\r\n',
});

const isConnect = await RNBluetoothClassic.isDeviceConnected(
  device.address,
);

if (!isConnect)
  throw new Error(`Failed to connect device ${device.name}`);
```

**Reply by kenjdavidson** [reply]
I'm still not sure why either:

- `connectToDevice` won't work, as it will either a) return a device or b) throw an exception if it cannot connect.

```
const device = await RNBluetoothClassic.connectToDevice(device.address, {
  connectionType: 'binary',
  delimiter: '\r\n',
});
```

- `connect` which would be done by using the connect method, which you've seen

```
const device = getBondedDevices().find(address -> address == "the address"); // or something else
const connected = device.connect();
if (connected).... 
else ....
```

As mentioned, the exception you're throwing is from attempting to `onDeviceRead` when the device is not connected.   So you should have either gotten a connection exception or a false from the two above methods.
